File: main/Graph.cpp
#include "Graph.hpp" 

Graph::Graph(){};
Graph::~Graph(){};

Graph::Graph( const Graph& ref){
	this->citys = ref.citys;
	this->citys_names = ref.citys_names;
	this->transports = ref.transports;
	this->transports_names = ref.transports_names;
	this->flights = this->flights;
};

id_t Graph::addCity(const std::string name){
	if (name.empty()){
		throw std::invalid_argument( "empty name");
	}
	// есть ключ в массиве, если нет, то FindCityByName кидает exp
	try
	{
		return this->findCityByName(name);
	}
	catch(std::domain_error e)
	{}
	//std::unordered_map<std::string, id_t>::iterator itr = this->citys.find(name);
	//if (itr != this->citys.end()){
	//	return *itr;
	//}
	//размер = свободное место	
	id_t nums = this->getNumCitys();

	this->citys.insert(std::make_pair(name, nums));
	this->citys_names.insert(std::make_pair(nums, name));
	location v;
	this->flights.push_back(v);
	
	return nums;
};

id_t Graph::addTransport(const std::string name){
	if (name.empty()){
		throw std::invalid_argument( "empty name");
	}
	// есть ключ в массиве
	try
	{
		return this->findTransportByName(name);
	}
	catch(std::domain_error e){}
	
	id_t nums = this->getNumTypeTransports();
	this->transports.insert(std::make_pair(name, nums));
	this->transports_names.insert(std::make_pair(nums, name));
	return nums;
};

//(id_t from, id_t to, id_t transport_type, id_t cruise_time, id_t cruise_fare,id_t locals
void Graph::addFlight(id_t from,
	id_t to, 
	id_t transport_type, 
	weight_t time, 
	weight_t fare, 
	weight_t locals
	){
	id_t nums =this->getNumCitys();
	if (nums < from || nums < to) {
		throw std::invalid_argument("there is no such city");
	}
	if (this->transports.size() < transport_type){
		throw std::invalid_argument("incorrect transport type");
	} 

	this->flights[from].push_back(Flight(from, to, transport_type, time, fare, locals));
};

// ищет город, если такого нет, то кидает исслючение
id_t Graph::findCityByName(const std::string name) {
	auto itr = this->citys.find(name);
	if (itr == this->citys.end()){
		throw std::domain_error("There is no such name."); 
	}
	return this->citys[name];
};
std::string Graph::findCityById(const id_t id){
	//std::unordered_map<std::string, id_t>::iterator itr = this->citys_names.find(id);
	auto itr = this->citys_names.find(id);
	if (itr == this->citys_names.end()){
		throw std::domain_error("There is no such id."); 
	}
	return this->citys_names[id];
};

id_t Graph::findTransportByName(const std::string name) {
	std::unordered_map<std::string, id_t>::iterator itr = this->transports.find(name);
	if (itr == this->transports.end()){
		throw std::domain_error("there is no such name."); 
	}
	return this->transports[name];	
};

std::string Graph::findTransportById(const id_t id) {
	//std::unordered_map<std::string, id_t>::iterator itr = this->transports_names.find(id);
	auto itr = this->transports_names.find(id);
	if (itr == this->transports_names.end()){
		throw std::domain_error("There is no such id."); 
	}
	return this->transports_names[id];
};

id_t Graph::getNumCitys() const{
	return this->flights.size();
};

id_t Graph::getNumTypeTransports() const{
	return this->transports.size(); 	
};
 
std::string Graph::flightsToString(const Flight& flight){ 
	std::string from = this->citys_names[ flight.get_from()];
	std::string to = this->citys_names[flight.get_to()];
	std::string transport = this->transports_names[flight.get_transport()];
	weights_t w = flight.get_weights();
	std::string result = from + " -> " + to + " тр-рт:" + transport + " время:" + std::to_string(w[CRUISE_TIME_P]) + " цена:" + std::to_string(w[CRUISE_FARE_P]);  
	//std::format("{} -> {} тр-рт{} время:{} цена:{} \n", from, to, transport, w[CRUISE_TIME_P], w[CRUISE_FARE_P]);
	return result + "\n";
};
//вывод в файл, надо сделать оператор <<
void Graph::printInFile(std::ofstream& out){
	if ( ! out.is_open()){
		throw std::runtime_error("Файл не открыт\n");
	}
	for( location loc :this->flights ){
		for(Flight f :loc){
			out<<(this->flightsToString(f)); 
		}
	}
	out<<std::endl;
};
id_t Graph::size() const{
	return this->getNumCitys();
};

File: py/graph.py
from math import fabs
from py.flight import Flight

#ИМЕНА
#location = list[]std::vector<Flight>;
MAX_ID = 2**32
MAX_WEIGHT = 2**32
'''/*class Location{
protected:
	id_t id;
	vector<Flight> flights; //нужен лю using?
public:
 
	Location(id_t id);
	~Location();
};
*/
'''

class Graph:
	def __init__(self):
		# [string, id]
		self.citys = {}
		#<id_t, std::string>
		self.citys_names = {}
		#<std::string, id_t>
		self.transports_name_id = {}
		#<id_t, std::string>
		self.transports_id_name = {}
		#vector<Flight>
		self.flights = {}
		self.num_citys = 0
		self.num_type_transports = 0
	@property
	def num_citys_upd(self):
		""" id_t getNumCitys() const;"""
		self.num_citys = len(self.citys)
		return self.num_citys  
		pass	
	@property
	def num_type_transports_upd(self):
		self.num_type_transports = len(self.transports_name_id)
		return self.num_type_transports  
	
	def size(self):
		"""id_t size() const;"""
		return self.num_citys_upd
	def __getitem__(self, index: int):
		"""location& operator[] (int i)"""
		if fabs(index) > self.num_citys:
			raise ValueError( f"index:{index} > MAX_ID:{MAX_ID}")
		poss = index if index >= 0 else (self.num_citys + index)   
		return self.flights[poss]
		pass
	def add_city(self, name: str):	
		"""id_t addCity(const std::string name);"""
		if not name: 
			raise ValueError( "empty name")
		if name in self.citys_names:
			return self.citys_names[name]
		nums  = self.num_citys
		self.citys[nums] = name
		self.citys_names[name] = nums 
		self.flights[nums] = []
		self.num_citys = self.num_citys + 1
		return nums
	def add_transport(self, name: str):
		"""id_t addTransport(const std::string name);"""
		if not name: 
			raise ValueError( "empty name")
		if name in self.transports_name_id:
			return self.transports_name_id[name]
		nums = self.num_type_transports
		self.transports_name_id[name] = nums
		self.transports_id_name[nums] = name 
		self.num_type_transports = self.num_type_transports + 1 
		return nums

	def add_flight(self, 
		from_city: int, to_city: int,
		transport_type: int, 
		time = MAX_WEIGHT, 
		fare = MAX_WEIGHT, 
		locals_w = 1
		):
		"""	void addFlight(id_t from,
		id_t to, 
		id_t transport_type, 
		weight_t time, 
		weight_t fare, 
		weight_t locals
		);
		"""
		if self.num_citys < from_city or self.num_citys < to_city: 
			raise ValueError("there is no such city")
		if self.num_type_transports < transport_type:
			raise ValueError("incorrect transport type")
		self.flights[from_city].append(Flight(from_city, to_city, transport_type, time, fare, locals_w))

	def find_city_by_name(self, name: str):
		"""id_t find_city_by_name(const std::string name);"""
		if name in self.citys_names:
			return self.citys_names[name]
		return -1
		raise ValueError("there is no such city")
	def find_transport_by_name(self, name: str):
		"""id_t findTransportByName(const std::string name);"""
		if name in self.transports_name_id:
			return self.transports_name_id[name]
		return -1
		raise ValueError("there is no such transport")
	def find_city_by_id(self, id_city: int):
		"""std::string findCityById(const id_t id);"""
		if id_city < self.num_citys:
			return self.citys[id_city]
		return -1
		raise ValueError("there is no such city")
		pass
	def find_transport_by_id(self, id_tr: int):
		"""std::string findTransportById(const id_t id);"""
		if id_tr < self.num_type_transports:
			return self.transports_id_name[id_tr]
		raise ValueError("there is no such Transport By Id")
		pass
	def flight_to_string(self, flight: Flight):
		"""std::string flightToString(const Flight& flight);"""
		return f"{self.find_city_by_id(flight.from_city)} -> {self.find_city_by_id(flight.to_city)}\
		{self.find_transport_by_id(flight.transport_type)} {flight.weights}"
		pass
	def printInFile(self, out):
		"""void printInFile(std::ofstream& out);"""
		pass
	



#endif /* GRAPH_H */
